### PR TITLE
Add rnn args check

### LINKED ISSUE #3851: Improve error message for RuntimeError: inconsistent tensor size in RNN
I run RNN sucessfully with the code.
```
rnn = nn.RNN(features, hidden_size, num_layers)
o, hidden = rnn(x, hidden)
```
and I run embedding example. Then, got RuntimeError.
```
RuntimeError: inconsistent tensor size, expected r_ [5 x 25], t [5 x 25] and src [2 x 25] to 
have thesame number of elements, but got 125, 125 and 50 elements respectively at 
/opt/conda/conda-bld/pytorch_1503970438496/work/torch/lib/TH/generic/THTensorMath.c:887
```
I set `batch_first=True` to use `Embedding`
```
x.size is [5, 8, 15]
hidden_size is [5, 2, 25]
```
nothing seems wrong.
And the error info seems not the dim problem?
The error is so strange.

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #3851, #3259

Added a high-level check for arguments to RNNBase (these were moved from arg checks in `cudnn/rnn.py`)

### Test Plan
`python test/test_nn.py`

### CODE REVIEW

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -128,6 +128,33 @@ def reset_parameters(self):
...
+            mini_batch = batch_sizes[0]
+        else:
+            mini_batch = input.size(0) if self.batch_first else input.size(1)
+
+        num_directions = 2 if self.bidirectional else 1
+        expected_hidden_size = (self.num_layers * num_directions,
+                                mini_batch, self.hidden_size)
+        if self.mode == 'LSTM':
+            hidden = hidden[0]
+        if tuple(hidden.size()) != expected_hidden_size:
```
Comment (apaszke) [inline torch/nn/modules/rnn.py]: You should test both in case of LSTM (they have the same size)

**Review by apaszke — COMMENTED** [review]
Looks good, but please check everything for LSTM.

**Review by apaszke — APPROVED** [review]

**Review by apaszke — APPROVED** [review]
Didn't this PR have a test for these cases as well? It would be nice to add it back before merging.

**Review by apaszke — APPROVED** [review]

**Comment by Shandilya21** [comment]
@zou3519  I have similar problem, 

RuntimeError: Expected hidden[0] size (1, 64, 256), got (64, 256),  i  tried different ways but unable to get it.
here is the code snippet.:
def forward(self, input_src, input_trg, courteous_template, ctx_mask=None, trg_mask=None):
        src_emb = self.src_embedding(input_src)
        trg_emb = self.trg_embedding(input_trg)
        temp_emb = self.temp_embedding(courteous_template)


        self.h0_encoder, self.c0_encoder = self.get_state(input_src)
        self.h1_encoder, self.c1_encoder = self.get_courteous(courteous_template)

        src_h, (src_h_t, src_c_t) = self.encoder(
            src_emb, (self.h0_encoder, self.c0_encoder)
        )
        

        temp_h, (tmp_h_t, tmp_c_t) = self.temp_encoder(
            temp_emb, (self.h1_encoder, self.c1_encoder)
        )

        out = torch.cat((src_h, temp_h),1)


        out = out.reshape(out.size(1),
        	out.size(0),out.size(2))

        # print(out.size())

        h_t = out[-1]
        h = h_t.view(h_t.size(0),h_t.size(1))



        trg_h, (_,_) = self.decoder(
        	trg_emb, h_t.view(
        		h_t.size(0),
        		h_t.size(1))
        	)
   RuntimeError: Expected hidden[0] size (1, 64, 256), got (64, 256),  i  tried different ways but unable to get it.
please help to rectify this issues

**Comment by zou3519** [comment]
@Shandilya21 Please ask a question on the forums or open an issue if you think there is a bug.

**Pull request merged**